### PR TITLE
WT-12619 Refactor WT_REF state setting and accessing

### PR DESCRIPTION
**Pull request by luke-pearson**
This standardizes code accessing the ref state and provides a minimal API for operations involving the ref state:
- It wraps reading ref state in a `WT_GET_REF_STATE` which allows us to: bury `ref->state` under `__` so it's `ref->__state`.
- Deleted the `WT_REF_SAVE_STATE` macro and replaced it with a function

### CODE REVIEW

File: src/btree/bt_random.c
Comment (luke-pearson) [inline src/btree/bt_random.c:412]: I figure as we're doing a volatile atomic load this should be fine.

File: src/btree/bt_sync_obsolete.c
Comment (agorrod) [inline src/btree/bt_sync_obsolete.c:219]: If we take this change the comment should be updated - it is no longer true (I presume)

**Review by agorrod — COMMENTED** [review]
I'm generally in favour of this change - it cleans things up and makes it clear that the ref state shouldn't be messed with directly (which I'm a fan of).

I'm concerned about the performance fallout - we'd need to measure (and possibly justify) that.

I'm also a little worried that it feels like "locking" - reading the ref state locally guarantees that we saw a ref that has been in that state, but not that it continues to be in that state. The code remains subtle in that way - it's probably better as-is, but I've known the code for so long I don't have good perspective.

File: src/btree/bt_random.c
Comment (ajmorton) [reply src/btree/bt_random.c:412]: I'm mostly agreed. Atomic volatile reads have the same effect as READ_ONCE, and while I'd prefer READ_ONCE for clarity I think making it work with ref->__state is going to have a few annoying implementation details. Happy to keep this as is.

**Review by ajmorton — COMMENTED** [review]
I'm picking up this PR while Luke is on leave. I've left a few comments for myself to touch up.

**Comment by ajmorton** [comment]
> I'm concerned about the performance fallout - we'd need to measure (and possibly justify) that.

SGTM. I didn't see any perf impacts before we split this work out of WT-12589, but I'll re-run perf testing to confirm.

> I'm also a little worried that it feels like "locking" - reading the ref state locally guarantees that we saw a ref that has been in that state, but not that it continues to be in that state. The code remains subtle in that way - it's probably better as-is, but I've known the code for so long I don't have good perspective.

I have a preference for local variables since an out-of-date-but-stable `ref->state` is easier to debug than one that could change part way through the function and start taking different code paths. However to be conservative we've limited the addition of local variables to only when it's (in my opinion) obvious that was the intent like `(state == WT_REF_DISK || state == WT_REF_MEM)` checks. 
For cases where it's not obvious, like the two reads of `ref->state` in `__wt_delete_page`, we've kept multiple reads of ref->state to minimise behavioural changes in the PR. I'll do a second walk through to confirm this is the case.

File: src/include/ref_inline.h
Comment (ajmorton) [reply src/include/ref_inline.h]: [758242d](https://github.com/wiredtiger/wiredtiger/pull/10345/commits/758242d2b41eea509e1e9102ff5517c832cd5910)

File: src/btree/bt_split.c
Comment (ajmorton) [reply src/btree/bt_split.c]: [d645942](https://github.com/wiredtiger/wiredtiger/pull/10345/commits/d6459428adcbc148c1483db18e1e2170daff61ae)

File: src/include/ref_inline.h
Comment (ajmorton) [reply src/include/ref_inline.h:1]: Closing this out. We have a few examples of `*_inline.h` files that aren't paired with a `*.h` file in `src/include`. I think this gives us the best of both worlds and we can always make future changes in a smaller PR.

File: src/btree/bt_delete.c
Comment (luke-pearson) [inline src/btree/bt_delete.c:134]: This switch case is uneeded complexity for what should be an `if (__wt_ref_get_state(ref) != WT_REF_DISK) return (0);`

Comment (ajmorton) [reply src/btree/bt_delete.c:134]: Updated in [d2c93e2](https://github.com/wiredtiger/wiredtiger/pull/10345/commits/d2c93e207197226e60cacfaf2108061f1058f105). I've kept `previous_state = __wt_ref_get_state(ref);` since we also use `previous_state` in the following cas.

File: src/include/btmem.h
Comment (luke-pearson) [inline src/include/btmem.h:1091]: This comment is good but needs some fixing up:
```
  /*
     * Ref state: Obscure the field name as this field shouldn't be accessed directly. The public
     * interface is made up of five functions:
     *  - __wt_ref_get_state 
     *  - WT_REF_SET_STATE
     *  - WT_REF_CAS_STATE
     *  - __wt_ref_lock
     *  - WT_REF_UNLOCK
     *
     * For more details on these functions see ref_inline.h.
     */
```

File: src/include/ref_inline.h
Comment (luke-pearson) [inline src/include/ref_inline.h:20]: Add a comment here:
```
/*
 * # The ref state API. #
 *
 * 5 functions* (3 macros) are defined to manipulate the ref state. This is a highly sensitive
 * field and protected via the double underscore keyword. Ideally no one will access it without
 * using these functions. Sadly being C we can't do much better than this.
 *
 * __wt_ref_get_state:
 * Get the state of the ref, wraps a relaxed atomic volatile load. At the time of writing this
 * comment this was done to enable TSAN and to enable burying the field behind the 
 * aforementioned double underscore.
 *
 * WT_REF_SET_STATE:
 * Set the ref state. If HAVE_REF_TRACK is defined track where the set call originated from. The
 * ref state tracking is why we use a macro here, as it utilizes gcc identifiers to track the
 * function and line number from where the macro was called.
 *
 * WT_REF_CAS_STATE:
 * Swap in a new state to the ref, again track where the call originated from.
 *
 * __wt_ref_lock:
 * Spin until the state WT_REF_LOCKED is swapped into the ref state field. Once the call to this
 * function completes the caller has exclusive access to the ref.
 *
 * WT_REF_UNLOCK:
 * Effectively wraps WT_REF_SET_STATE however should only be used when returning the ref to the
 * previous state as returned by __wt_ref_lock.
 */
```

File: src/include/ref_inline.h
Comment (ajmorton) [reply src/include/ref_inline.h]: Yeah I changed it since we had calls to the function outside of ref_inline.h and pattern matched on our style guide.
I've rolled this back in [e21f8c0](https://github.com/wiredtiger/wiredtiger/pull/10345/commits/e21f8c011ebd1f98f34215782f59dece563afacc) and changed those external calls to use the macro instead.

File: src/include/btmem.h
Comment (ajmorton) [reply src/include/btmem.h]: Updated in [1717dbe](https://github.com/wiredtiger/wiredtiger/pull/10345/commits/1717dbe030dab0ddbb2e7ab981c4e95e6a65cfd0).
I don't want to do so in this PR, but should `__wt_ref_lock` also become a macro? It calls `WT_REF_CAS_STATE` which uses the `__LINE__` and `__FUNCTION__` macros, so we're currently not reporting the ref_lock call site in our ref tracking

File: src/include/ref_inline.h
Comment (ajmorton) [reply src/include/ref_inline.h:20]: Added in [1717dbe](https://github.com/wiredtiger/wiredtiger/pull/10345/commits/1717dbe030dab0ddbb2e7ab981c4e95e6a65cfd0). I've made some small tweaks so please review the new comment

File: src/include/btmem.h
Comment (luke-pearson) [reply src/include/btmem.h]: I think its fine as is, at the very least a user of tracking will see that the ref went through one of the places that called `__wt_ref_lock`. Personally I don't think ref tracking get used that much so we shouldn't over index on it.

**Review by luke-pearson — COMMENTED** [review]
LGTM. I can't approve as its my PR tho.

File: src/include/ref_inline.h
Comment (marcbutler) [inline src/include/ref_inline.h:65]: Great comment. Is it possible to clarify why "we allow it to race" is it because it is diagnostic code or for another reason?

File: src/include/ref_inline.h
Comment (marcbutler) [inline src/include/ref_inline.h:147]: It's a small thing but I would ask you to consider typographically aligning lock/unlock. It seems easiest to my mind to wrap `__wt_ref_lock` in a straight `WT_REF_LOCK` macro.
This serves no purpose other than to aid developer cognition.

**Review by marcbutler — APPROVED** [review]
I tried to concentrate my efforts on `ref_inline.h`.

My review isn't as thorough as perhaps it could be -- I struggle with the fact that we fact we don't lock to maintain state as @agorrod has pointed out. But I also feel that this structuring of the code precedes this work, and this is not the PR to explore or address such concerns.

As discussed I think this will dovetail well with future plans. Thanks.

My comments are small considerations only so happy to give LGTM.

File: src/include/ref_inline.h
Comment (ajmorton) [reply src/include/ref_inline.h:65]: I believe it's because this is diagnostic code and some fuzziness in ref history is preferable to taking a lock and significant slowdowns. This was taken from the [original comment here](https://github.com/wiredtiger/wiredtiger/blob/3b8eb799ce99626cb8a827b0a8855440ba00fb24/src/include/btree_inline.h#L40).

We'll update this to `This is diagnostic code and ref->state changes are a hot path. As such we allow some racing in the history tracking code instead of requiring a lock and slowing down ref state transitions.`

File: src/include/ref_inline.h
Comment (luke-pearson) [reply src/include/ref_inline.h:150]: [a6c9128](https://github.com/wiredtiger/wiredtiger/pull/10345/commits/a6c91285f978ada0f1e95ebee0001b758e4515a1)

File: src/include/ref_inline.h
Comment (luke-pearson) [reply src/include/ref_inline.h]: 172bf5091

File: src/include/ref_inline.h
Comment (ajmorton) [reply src/include/ref_inline.h:98]: Yeah I'd prefer `#define WT_REF_GET_STATE(ref) __ref_get_state(ref)`. No harm in making it explicit

Comment (luke-pearson) [reply src/include/ref_inline.h]: [5cc8ba6](https://github.com/wiredtiger/wiredtiger/pull/10345/commits/5cc8ba6c2727950b91339483c7d4ebce226a859e)

**Review by ajmorton — APPROVED** [review]
LGTM, but I'm co-author so that doesn't carry much weight.
We'll wait for approval from @quchenhao before merging.

File: src/include/ref_inline.h
Comment (quchenhao) [inline src/include/ref_inline.h:95]: This is now always sequential consistency? It is not required to be that strict in most places.

Comment (luke-pearson) [reply src/include/ref_inline.h:95]: It's relaxed

**Review by quchenhao — APPROVED** [review]
lgtm

**Pull request merged**
